lab/lablib/midi/parse_midi.js

```diff
@@ -0,0 +1,279 @@
+ * @param {DataView} payload The binary data for the track chunk
+ * @param {number} offset The first byte of the variable length quantity
+ * @returns {[number, number]} (value, next_offset)
+ */
+function parse_variable_length(payload, offset) {
+  let index = offset;
+  let more_bytes = true;
+  let value = 0;
+  while (more_bytes) {
+    const byte = payload.getUint8(index);
+    // the low 7 bits are tacked on to the value (MSBF)
+    value = (value << 7) | (byte & 0x7f);
+
+    // the top bit of the byte indicates that there's more bytes to come.
+    more_bytes = byte >> 7 === 1;
+    index++;
+  }
+
+  return [value, index];
+}
+
+const META_MESSAGE = 0xff;
+const SYSEX_MESSAGE = 0xf0;
+const SYSEX_ESCAPE = 0xf7;
+
+const META_HEADER_LENGTH = 3;
+
+/**
+ * Parse a MIDI meta event
+ * @param {DataView} payload The payload of the track chunk
+ * @param {number} offset The offset for the start of the message (in this case, the FF byte)
+ * @param {number} tick_delta The time of the message
+ * @returns {[import("./MidiFile.js").MIDIMetaEvent, number]} (message, next_offset)
+ */
+function parse_meta_message(payload, offset, tick_delta) {
+  const meta_type = payload.getUint8(offset + 1);
+  const length = payload.getUint8(offset + 2);
+
+  let message;
+
+  if (meta_type == MIDIMetaType.END_OF_TRACK) {
+    message = new MIDIEndOfTrack(tick_delta);
+  } else if (meta_type === MIDIMetaType.TIME_SIGNATURE) {
+    const numerator = payload.getUint8(offset + 3);
+    const denominator = payload.getUint8(offset + 4);
+    const midi_clocks = payload.getUint8(offset + 5);
+    const notes32 = payload.getUint8(offset + 6);
+    message = new MIDITimeSignature(
+      tick_delta,
+      numerator,
+      denominator,
+      midi_clocks,
+      notes32
+    );
+  } else if (meta_type === MIDIMetaType.KEY_SIGNATURE) {
+    const sharps_flats = payload.getInt8(offset + 3);
+    const major_minor = payload.getInt8(offset + 4);
+    message = new MIDIKeySignature(tick_delta, sharps_flats, major_minor);
+  } else if (meta_type === MIDIMetaType.SET_TEMPO) {
+    const microseconds_per_quarter =
+      (payload.getUint8(offset + 3) << 16) |
+      (payload.getUint8(offset + 4) << 8) |
+      payload.getUint8(offset + 5);
+    message = new MIDISetTempo(tick_delta, microseconds_per_quarter);
+  } else {
+    const data = new Uint8Array(length);
+    for (let i = 0; i < length; i++) {
+      data[i] = payload.getUint8(offset + 3 + i);
+    }
+
+    console.warn(
+      "detected unsupported MIDI meta event",
+      "0x" + meta_type.toString(16)
+    );
+    message = new MIDIOtherMeta(tick_delta, meta_type, data);
+  }
+
+  return [message, offset + META_HEADER_LENGTH + length];
+}
+
+function parse_sysex_message(payload, offset, tick_delta) {
+  const [length, next_offset] = parse_variable_length(payload, offset + 1);
+
+  const data = new Uint8Array(length);
+  for (let i = 0; i < length; i++) {
+    data[i] = payload.getUint8(next_offset + i);
+  }
```

Comment:
It would be cleaner to go to the underlying buffer of `payload` and compute a new slice as a Uint8 array directly, no copying needed.